chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,600 @@
+		return nil, err
+	}
+
+	deals, err := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.Deals)
+	if err != nil {
+		// TODO: kind of annoying that this can be caused by gas, otherwise could be fatal
+		return nil, aerrors.HandleExternalError(err, "loading deals amt")
+	}
+
+	for _, deal := range params.Deals {
+		var dealInfo OnChainDeal
+		if err := deals.Get(deal, &dealInfo); err != nil {
+			return nil, aerrors.HandleExternalError(err, "getting del info failed")
+		}
+
+		workerBytes, err := vmctx.Send(dealInfo.Deal.Proposal.Provider, MAMethods.GetWorkerAddr, types.NewInt(0), nil)
+		if err != nil {
+			return nil, err
+		}
+		providerWorker, eerr := address.NewFromBytes(workerBytes)
+		if eerr != nil {
+			return nil, aerrors.HandleExternalError(eerr, "parsing provider worker address bytes")
+		}
+
+		if vmctx.Message().From != providerWorker {
+			return nil, aerrors.New(1, "ActivateStorageDeals can only be called by deal provider")
+		}
+
+		if vmctx.BlockHeight() > dealInfo.Deal.Proposal.ProposalExpiration {
+			return nil, aerrors.New(2, "deal cannot be activated: proposal expired")
+		}
+
+		if dealInfo.ActivationEpoch > 0 {
+			// this probably can't happen in practice
+			return nil, aerrors.New(3, "deal already active")
+		}
+
+		dealInfo.ActivationEpoch = vmctx.BlockHeight()
+
+		if err := deals.Set(deal, dealInfo); err != nil {
+			return nil, aerrors.HandleExternalError(err, "setting deal info in AMT failed")
+		}
+	}
+
+	dealsCid, err := deals.Flush()
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "saving deals AMT")
+	}
+
+	self.Deals = dealsCid
+
+	nroot, err := vmctx.Storage().Put(&self)
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "storing state failed")
+	}
+
+	aerr := vmctx.Storage().Commit(old, nroot)
+	if aerr != nil {
+		return nil, aerr
+	}
+
+	return nil, nil
+}
+
+type ProcessStorageDealsPaymentParams struct {
+	DealIDs []uint64
+}
+
+func (sma StorageMarketActor) ProcessStorageDealsPayment(act *types.Actor, vmctx types.VMContext, params *ProcessStorageDealsPaymentParams) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	deals, err := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.Deals)
+	if err != nil {
+		// TODO: kind of annoying that this can be caused by gas, otherwise could be fatal
+		return nil, aerrors.HandleExternalError(err, "loading deals amt")
+	}
+
+	for _, deal := range params.DealIDs {
+		var dealInfo OnChainDeal
+		if err := deals.Get(deal, &dealInfo); err != nil {
+			return nil, aerrors.HandleExternalError(err, "getting del info failed")
+		}
+
+		encoded, err := CreateExecParams(StorageMinerCodeCid, &IsMinerParam{
+			Addr: vmctx.Message().From,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		ret, err := vmctx.Send(StoragePowerAddress, SPAMethods.IsMiner, types.NewInt(0), encoded)
```

Comment:
We should pull this check out of the loop, doesnt need to be done per deal does it?